### PR TITLE
Implement websocket signaling (webapp/webclient)

### PR DESCRIPTION
**Pull request by kannan-xiao4**
unity client websocket implement using https://github.com/Unity-Technologies/UnityRenderStreaming/pull/209

## ToDo
- Server implementation is not yet done, working in progress.
- Currently, I don't operation check yet.

### CODE REVIEW

**Comment by karasusan** [comment]
Review it

File: WebApp/src/websocket.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                    case "disconnect":
+                        this.onDisConnect(ws, msg.data);
+                        break;
+                    case "offer":
+                        this.onOffer(ws, msg.data);
+                        break;
+                    case "answer":
+                        this.onAnswer(ws, msg.data);
+                        break;
+                    case "cadicate":
```
Comment (karasusan) [inline WebApp/src/websocket.ts]: `cadidate` 
`candidate`

File: WebApp/src/websocket.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        if (mapCandidates) {
+          const arrayCandidates = mapCandidates.get(connectionId);
+          for (const candidate of arrayCandidates) {
+            candidate.datetime = Date.now();
+          }
+        }
+        clients.forEach((_v, k) => k.send(JSON.stringify({from:connectionId, to:"", type:"answer", data:newAnswer})));
+    }
+
+    private onCadicate(ws: WebSocket, message: any){
```
Comment (sindharta) [inline WebApp/src/websocket.ts]: Like what @karasusan said, this should be `onCandidate` too

File: WebApp/src/websocket.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    private onConnect(ws: WebSocket){
+        const connectionId: string = uuid();
+        const connectionIds = getOrCreateConnectionIds(ws);
+        connectionIds.add(connectionId);
+        ws.send(JSON.stringify({connectionId:connectionId}));
+    }
+
+    private onDisConnect(ws: WebSocket, message: any){
```
Comment (sindharta) [inline WebApp/src/websocket.ts]: `onDisconnect` would be better.

File: WebApp/src/websocket.ts (hunk 1 above)
Comment (kannan-xiao4) [reply WebApp/src/websocket.ts]: I fixed it

File: WebApp/src/websocket.ts (hunk 2 above)
Comment (kannan-xiao4) [reply WebApp/src/websocket.ts]: I fixed it

File: WebApp/src/websocket.ts (hunk 3 above)
Comment (kannan-xiao4) [reply WebApp/src/websocket.ts]: I fixed it

File: WebApp/public/scripts/signaling.js [hunk 4]
```diff
@@ -1,54 +1,186 @@
...
+  async getCandidate(fromTime = 0) {
+    return await fetch(this.url(`candidate?fromtime=${fromTime}`), { method: 'GET', headers: this.headers() });
   };
 }
+
+export class WebSocketSignaling extends EventTarget {
+
+  constructor() {
+    super();
+    this.websocket = new WebSocket("ws://localhost:80");
```
[the remark this answers is not in the document]
Comment (kannan-xiao4) [reply WebApp/public/scripts/signaling.js]: ok, I will change to this URL can change from the. constructor argument.

[the remark this answers is not in the document]
Comment (kannan-xiao4) [reply WebApp/public/scripts/signaling.js]: This URL should be use location host.
I think we don't need change from the constructor argument.

[the remark this answers is not in the document]
Comment (karasusan) [reply WebApp/public/scripts/signaling.js]: OK

**Comment by karasusan** [comment]
LGTM

**Pull request merged**